source/common/http/http1/parser_impl.cc

```diff
@@ -0,0 +1,225 @@
+    llhttp_init(&parser_, type, &settings_);
+    llhttp_set_lenient_chunked_length(&parser_, 1);
+  }
+
+  Impl(llhttp_type_t type, void* data) : Impl(type) {
+    parser_.data = data;
+    settings_ = {
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onMessageBegin();
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onUrl(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        nullptr, // on_status
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderField(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderValue(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto statusor = conn_impl->onHeadersComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(statusor));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          static_cast<ParserCallbacks*>(parser->data)->bufferBody(at, length);
+          return 0;
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onMessageComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          // A 0-byte chunk header is used to signal the end of the chunked body.
+          // When this function is called, http-parser holds the size of the chunk in
+          // parser->content_length. See
+          // https://github.com/nodejs/http-parser/blob/v2.9.3/http_parser.h#L336
+          const bool is_final_chunk = (parser->content_length == 0);
+          static_cast<ParserCallbacks*>(parser->data)->onChunkHeader(is_final_chunk);
+          return 0;
+        },
+        nullptr, // on_chunk_complete
+        nullptr, // on_url_complete
+        nullptr, // on_status_complete
+        nullptr, // on_header_field_complete
+        nullptr  // on_header_value_complete
+    };
+  }
+
+  RcVal execute(const char* slice, int len) {
+    llhttp_errno_t error;
+    if (slice == nullptr || len == 0) {
+      error = llhttp_finish(&parser_);
```

Comment:
Can we detect the error condition were execute(non-null) is called after execute(null)?

I think this is mostly an issue in the codec interface, end_stream is not plumbed all the way; execute(null) should only happen when buffer is empty and end_stream is true.

Reply:
I could add internal state checking?
But we are guaranteed that the len condition will be OK since that's what the caller enforces, and it comes down to whether slice is not nullptr when len > 0.
https://github.com/envoyproxy/envoy/blob/9e263d8d636fb0b52da99a524b906cfc62190ae0/source/common/http/http1/codec_impl.cc#L577